[17.0][MIG] purchase_sale_stock_inter_company

cuongnmtm opened the pull request:
Depends on:

- [x] <https://github.com/OCA/multi-company/pull/595>
- [x] <https://github.com/OCA/multi-company/pull/604>

yankinmax committed:
[ADD][15.0] purchase_sale_stock_inter_company

JasminSForgeFlow committed:
[15.0][IMP] purchase_sale_stock_inter_company:
Update v14 latest changes and test cases

MiquelRForgeFlow committed:
[IMP] purchase_sale_stock_inter_company: reflect purpose of module

chafique-delli committed:
[IMP] purchase_sale_stock_inter_company: pre-commit stuff

chafique-delli committed:
[MIG] purchase_sale_stock_inter_company: Migration to 16.0

JordiBForgeFlow committed:
[FIX] purchase_sale_stock_inter_company: use reserved_qty instead of product_qty in stock.move.line

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (10 of 10 strings)

Translation: multi-company-16.0/multi-company-16.0-purchase_sale_stock_inter_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-16-0/multi-company-16-0-purchase_sale_stock_inter_company/es/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 90.0% (9 of 10 strings)

Translation: multi-company-16.0/multi-company-16.0-purchase_sale_stock_inter_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-16-0/multi-company-16-0-purchase_sale_stock_inter_company/it/

JordiMForgeFlow committed:
[IMP] purchase_sale_stock_inter_company: refactor stock picking validation logic for inheritance in purchase_sale_stock_inter_company_mrp

JordiMForgeFlow committed:
[ADD] purchase_sale_stock_inter_company_mrp

pomazanbohdan commented:
/ocabot migration purchase_sale_stock_inter_company

pomazanbohdan commented:
@cuongnmtm Would you be so kind as to repeat the note to the bot?
> /ocabot migration purchase_sale_stock_inter_company

cuongnmtm commented:
/ocabot migration purchase_sale_stock_inter_company

cuongnmtm commented:
I don't have permission, but I believe the maintainer will somehow receive a notification about this :D

Kev-Roche committed:
[16.0][FIX] add test: synchronize inter-company picking qty in case of backorder

Kev-Roche committed:
[16.0][FIX] add fix: synchronize inter-company picking qty in case of backorder

Kev-Roche committed:
[16.0][FIX][purchase_sale_stock_inter_company] multiple products leads to a backorder

carlos-lopez-tecnativa commented on purchase_sale_stock_inter_company/tests/test_inter_company_purchase_sale_stock.py:
```diff
@@ -5,10 +5,12 @@
 )
 
+TestPurchaseSaleInterCompany = test_icps.TestPurchaseSaleInterCompany
+
```
Why?

cuongnmtm replied:
Pre-commit auto fixes I guess

It just refactors the code.

carlos-lopez-tecnativa reviewed (changes requested):
Please rebase the branch
TT54931

carlos-lopez-tecnativa reviewed (changes requested):
Please include and adapt https://github.com/OCA/multi-company/pull/733

carlos-lopez-tecnativa commented:
> Please include and adapt #733

ping @cuongnmtm 

And remove the last commit, the dependency has already been merged.

cuongnmtm commented:
Hello @carlos-lopez-tecnativa , I plan to allocate time to work on this later this week. If it's urgent on your end, please feel free to continue from where I left off. Thank you.

pedrobaeza commented:
#804 should be included here.

cuongnmtm commented:
I cherry-picked the related commits from:
- https://github.com/OCA/multi-company/pull/572
- https://github.com/OCA/multi-company/pull/789
- https://github.com/OCA/multi-company/pull/804

and adapted the migration code with the refactored code.

cuongnmtm commented:
Something went wrong. I tried to solve conflicts, and the migration changes were moved to the conflicted commit. I will need to redo it.

namtnt committed:
[IMP] purchase_sale_stock_inter_company: pre-commit auto fixes

namtnt committed:
[MIG] purchase_sale_stock_inter_company: Migration to 17.0

This commit includes not only the module migration but also additional logic to handle the synchronization of stock.move.line quantities and lots, inspired by this commit https://github.com/OCA/multi-company/pull/733/commits/511840885842ef7f92bcfa613cde633950d3e5a3 from the OCA multi-company repository.

Although best practices suggest that a migration commit should focus solely on the migration, separating the sync logic into a different commit before or after the migration was not appropriate in this case, as the migration alone does not address all the necessary scenarios.

cuongnmtm commented:
With the help of my teammate @namtnt , 

We added the improvement of 

https://github.com/OCA/multi-company/pull/572
https://github.com/OCA/multi-company/pull/789
https://github.com/OCA/multi-company/pull/804

and also adapt https://github.com/OCA/multi-company/pull/733 to have sync qty and lot.

metaminux commented on purchase_sale_stock_inter_company/tests/test_inter_company_purchase_sale_stock.py:
```diff
@@ -0,0 +1,440 @@
+                },
+            ),
+        ]
+        so_picking_id.sudo().button_validate()
```
Test is failing here, because we try to deliver only 1 of the 3 requested products and button_validate is opening the dialog to create a backorder...

```suggestion
        res_dict = so_picking_id.sudo().button_validate()
        if isinstance(res_dict, dict) and "context" in res_dict:
            self.env["stock.backorder.confirmation"].with_context(
                **res_dict["context"]
            ).process()
```
I think using the same code as in previous tests with backorder creation would make this test pass.

metaminux commented on purchase_sale_stock_inter_company/tests/test_inter_company_purchase_sale_stock.py:
```diff
@@ -0,0 +1,440 @@
+                },
+            ),
+        ]
+        so_picking_id.sudo().button_validate()
```
Test is failing here, because we try to deliver only 30 of the 50 requested "stockable_product" and button_validate is opening the dialog to create a backorder...

```suggestion
        res_dict = so_picking_id.sudo().button_validate()
        if isinstance(res_dict, dict) and "context" in res_dict:
            self.env["stock.backorder.confirmation"].with_context(
                **res_dict["context"]
            ).process()
```
I think using the same code as in previous tests with backorder creation would make this test pass.

metaminux commented:
Hi @cuongnmtm 

Can you take a look at my suggestions ? I think this could unlock tests here and help moving forward for other PRs that depend on this one...

Thanks

cuongnmtm commented:
@metaminux  Thanks for your suggestion. I am looking at it now.

cuongnmtm commented:
There are two issues when process the backorder wizard:
- test_sync_picking_lot_and_qty_with_move_diff_2:  You need to supply a Lot/Serial Number for product:  Stockable Product Tracked by Serial
- test_inter_company_purchase_sale_stock: This lot 111 is incompatible with this product Stockable Product


I am checking it

cuongnmtm commented:
Need to remove the extra moves in case:
```python
# remove the extra move lines in the receipt of lot tracking product
# example: In the receipt, we have 3 move lines for 3 different serials,
# in the delivery we specify 2 serials. When validating the delivery and
# creating back order, Odoo generates 3 move lines in the receipt, so
# we need to remove 1 different move line in the receipt, otherwise it
# will cause an error saying that we need to assign a lot or serial
# for the remaining move line
```

carlos-lopez-tecnativa commented on purchase_sale_stock_inter_company/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,7 @@
+This module considers the propagation of the warehouse into the sale
+orders.
+
```
I think this description should be improved, this module does more than that.

carlos-lopez-tecnativa commented on purchase_sale_stock_inter_company/tests/test_inter_company_purchase_sale_stock.py:
```diff
@@ -0,0 +1,396 @@
+                            "lot_id": self.serial_1.id,
+                            "picking_id": so_picking_id.id,
+                        },
+                    ),
```
Please use [Command.create](https://github.com/odoo/odoo/blob/9748b45b24140416fc6e0e87c25c60285bef3d70/odoo/fields.py#L4259); it's more readable.

```suggestion
                    Command.create(
                        {
                            "location_id": move.location_id.id,
                            "location_dest_id": move.location_dest_id.id,
                            "product_id": self.stockable_product_serial.id,
                            "quantity": 1,
                            "lot_id": self.serial_1.id,
                            "picking_id": so_picking_id.id,
                        },
                    ),
```

carlos-lopez-tecnativa commented on purchase_sale_stock_inter_company/tests/test_inter_company_purchase_sale_stock.py:
```diff
@@ -0,0 +1,396 @@
+                    (
+                        0,
+                        0,
+                        {
```
Here and in the other occurrences

carlos-lopez-tecnativa reviewed (changes requested):
When I try to validate a picking from the sale order, I get this error:
![image](https://github.com/user-attachments/assets/7c28dc8c-42f4-4673-a4b3-d6ed8c0ca96e)

cuongnmtm commented:
> When I try to validate a picking from the sale order, I get this error: ![image](https://private-user-images.githubusercontent.com/174346082/457370828-7c28dc8c-42f4-4673-a4b3-d6ed8c0ca96e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTA0MjU2MTcsIm5iZiI6MTc1MDQyNTMxNywicGF0aCI6Ii8xNzQzNDYwODIvNDU3MzcwODI4LTdjMjhkYzhjLTQyZjQtNDY3My1hNGIzLWQ2ZWQ4YzBjYTk2ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYyMFQxMzE1MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OTI2ZmVjYzcxY2Y4NGE5ZDQxZWYwNGZkNDg1MGQyNzllZmVlZDI0OGE3NDlmZjM4NTg0MzkzY2Q5Njg1MzVjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.YW8VYr4Hvya_-CIUG7Z_YBX5rXQQJ8SNy2EplYzvpOY)

This is multi-company rule. I will add sudo to bypass it.

cuongnmtm commented:
> > When I try to validate a picking from the sale order, I get this error: ![image](https://private-user-images.githubusercontent.com/174346082/457370828-7c28dc8c-42f4-4673-a4b3-d6ed8c0ca96e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTA0MjU2MTcsIm5iZiI6MTc1MDQyNTMxNywicGF0aCI6Ii8xNzQzNDYwODIvNDU3MzcwODI4LTdjMjhkYzhjLTQyZjQtNDY3My1hNGIzLWQ2ZWQ4YzBjYTk2ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYyMFQxMzE1MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OTI2ZmVjYzcxY2Y4NGE5ZDQxZWYwNGZkNDg1MGQyNzllZmVlZDI0OGE3NDlmZjM4NTg0MzkzY2Q5Njg1MzVjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.YW8VYr4Hvya_-CIUG7Z_YBX5rXQQJ8SNy2EplYzvpOY)
> 
> This is multi-company rule. I will add sudo to bypass it.

actually no need sudo but with_company does the job

cuongnmtm commented:
Thanks @carlos-lopez-tecnativa  for your review.
I updated the code to add a missing with_company to solve your issue, and updated these tests, which hide this issue.
I updated the readme and test as well.

carlos-lopez-tecnativa commented:
> Please include and adapt #733

From that PR, I noticed that this commit https://github.com/OCA/multi-company/pull/733/commits/bc478bf37e06cc9a2ad38c86c6bffaca0bdf80ca which synchronizes `stock.move.line` when the move is in the done state (and the picking is unlocked), is not included.
Could you please include it or explain why it was discarded?

Additionally, this module has a strange history. In v14, it existed as a single module: `purchase_sale_inter_company`.
At Tecnativa, we backported it to v13 and split it into two modules: `purchase_sale_inter_company` and `purchase_sale_stock_inter_company`, as was done in v15, see this PR: https://github.com/OCA/multi-company/pull/547. In v14, I found these two PRs: https://github.com/OCA/multi-company/pull/600 and https://github.com/OCA/multi-company/pull/656 They were never included in v15, but I think their features could be useful here.
Could you please cherry-pick and adapt those PRs?

carlos-lopez-tecnativa reviewed (changes requested):
Additionally, I noticed that commit https://github.com/OCA/multi-company/pull/605/commits/8e705aef46a6efc349553ab4bcd9a3ce314a4825 (for pre-commit autofixes) should only contain automatic formatting changes. However, it includes functional/code changes that shouldn't be in that type of commit.

These extra changes should be moved to the migration commit in order to keep a clean and accurate Git history.

![image](https://github.com/user-attachments/assets/1c4d4a1b-fdbf-464f-8127-4e82a01653ea)

cuongnmtm commented:
@carlos-lopez-tecnativa 

> From that PR, I noticed that this commit [bc478bf](https://github.com/OCA/multi-company/commit/bc478bf37e06cc9a2ad38c86c6bffaca0bdf80ca) which synchronizes `stock.move.line` when the move is in the done state (and the picking is unlocked), is not included.
> Could you please include it or explain why it was discarded?

The logic is handled in stock.picking. All tests are passed.

> Additionally, this module has a strange history. In v14, it existed as a single module: purchase_sale_inter_company.
At Tecnativa, we backported it to v13 and split it into two modules: purchase_sale_inter_company and purchase_sale_stock_inter_company, as was done in v15, see this PR: https://github.com/OCA/multi-company/pull/547. In v14, I found these two PRs: https://github.com/OCA/multi-company/pull/600 and https://github.com/OCA/multi-company/pull/656 They were never included in v15, but I think their features could be useful here.
Could you please cherry-pick and adapt those PRs?

I will have a look on Friday

metaminux commented:
Hello @cuongnmtm and @carlos-lopez-tecnativa 

As mentioned in previous comments, the history of the module is quite a mess...

There were only one module **purchase_sale_inter_company** in 14.0 with maintainers : @aleuffre and @renda-dev 

It has been split in **purchase_sale_inter_company** and **purchase_sale_stock_inter_company** from 14.0 to 15.0 (see #370 and #371 by @JasminSForgeFlow)

But a lot of improvements were added to 14.0 after that and never ported to 15.0 because it's harder to forward port due to the split.
Some attempts were done, but it's not so easy (see sunflowerit#3 and a lot of comments in PRs when @pedrobaeza asked for porting to newer versions)

Then, original module was split in 13.0 also...

And now, some PRs were merged to 16.0, adding features already present in 14.0, and facing same problems with different ways to fix...

I'm trying to migrate modules **purchase_sale_stock_inter_company** to 18.0 and **purchase_sale_stock_inter_company_mrp** to 17.0 and 18.0...

I'd like to use this PR has base for the migrations, but I feel @cuongnmtm is a bit lonely here, right ?

@OCA/intercompany-maintainers Can we gather efforts to clean all that mess ?

I'm trying to get the full history of issues and PRs for each version from 13.0 to 16.0, i'll try to post it here (or in another PR...)
I think 14.0 is the one with all features and tests but only one module, maybe we need to rewrite modules in 16.0 before doing the migration ?

Any thoughts everyone ?

pedrobaeza commented:
I understand your pain, and there has been conflicting interests with some people only working on 14.0 while others working on upper versions. We can try to gather everything together through @carlos-lopez-tecnativa if you agree.

metaminux commented:
> I understand your pain, and there has been conflicting interests with some people only working on 14.0 while others working on upper versions. We can try to gather everything together through @carlos-lopez-tecnativa if you agree.

Sure, what would be the better way ?

I think porting tests from 14.0 to 16.0 one by one with code making each test pass in each commit would be easier for review...
Maybe opening a new issue with sub issues for each commit ?

carlos-lopez-tecnativa commented:
> > I understand your pain, and there has been conflicting interests with some people only working on 14.0 while others working on upper versions. We can try to gather everything together through @carlos-lopez-tecnativa if you agree.
> 
> Sure, what would be the better way ?
> 
> I think porting tests from 14.0 to 16.0 one by one with code making each test pass in each commit would be easier for review... Maybe opening a new issue with sub issues for each commit ?


I'm working on this PR to forward-port all features from V13 and V14. Please review this PR and let me know if you have any questions.

Superseded by https://github.com/OCA/multi-company/pull/856

pedrobaeza commented:
Let's close this one then. Do you agree, @cuongnmtm ?
